### ISSUE TITLE
Improve placeholders and variables in the model config

### ISSUE THREAD

**Issue by tristanheywood** [nicolas-chaulet/deeppointcloud-benchmarks#46]
Right now the only placeholder supported is 'FEAT', the number of input features. This placeholder needs to be resolved by each model separately. 

It would be better to resolve all placeholders as part of the core of the framework, so that model implementations don't need to worry about it. We could also allow users to do arithmetic in any of the fields (.e.g nn: [64, 64*2]). 

We could also have a 'define_placeholders' section of the config where users can create their own placeholders. E.g. 

```
MyModel:
    type: RSConv
    define_placeholders:
        L1_OUT: 64
    down_conv:
         nn: [[FEAT, 32, L1_OUT], [L1_OUT, L1_OUT*2, 64]
```

**Comment by nicolas-chaulet** [comment]
Even if the idea is quite appealing I am not 100% convinced that placeholders add something or simplifies the configuration of models for the following reasons:

- For inputs, the number of features is actually completely fixed by the dataset and the selected  model (nothing the user can control). For example for PoinNet++ it is `3 + number_features` but for KPConv it is `number_features`. Size of input layers is not configurable, that's why I liked the -1 approach that then gets replaced by each model independently. 

- When you have sequences of MLP it is a bit the same, in the example above the input to the second MLP (L1_OUT) is totally defined by the output of the first MLP, they have to match anyway. So again, not something that the user has any control over. The size of the second layer being 2*L1_OUT would add some value though. But not sure it is worth maintaining a whole system of regular expressions for that. I think it's ok to write your MLPs every time entirely

Bottom line I am not convinced it will help users while certainly adding complexity when first getting into the framework. Shall put it aside until we have some more people weighing in and see where it goes?

**Comment by tristanheywood** [comment]
You raise some good points, but

- The user can control which dataset they use with a given model, so they need some way of having the same model deal with different numbers of features. 'FEAT' is determined by the dataset, so for pointnet++ you'd write "nn: [3 + FEAT, ...]" and for KPConv "nn: [FEAT, ...]". I think it's good to make the config as transparent as possible. I found it very confusing when I specified one number for a MLP in the model, but the model printout showed a different number. 
- It wouldn't be implemented using regular expressions. The method I used on the RSConv branch was to call eval() on the string, with 'FEAT' as a local variable. This method allows for any arithmetic operations involving both placeholders and number literals and avoids complicated parsing logic. 

I do agree that it adds some complexity, but I think users would find it intuitive, since it would make the config act more like writing a dict in python.

**Comment by tchaton** [comment]
Hey @tristanheywood @nicolas-chaulet,

So if I understand properly, you want to hide the regular expression execution and add more placeholders.

I was thinking to add an Expression Resolver before the model creation. 
It would take the dataset nbfeatures, look for every text value (by recursively navigating in nested list)
and resolve any of them.
By doing, so we would remove errors coming from arguments not being resolved during model creation and the possibility to add new features as the one you are suggesting.

If you both agree, @tristanheywood do you want to take care of that ?

**Comment by tristanheywood** [comment]
@tchaton we can do this without any regex. An expression resolver sounds like a good idea. 

I'll start working on it so we have a better idea of what it looks like. Then we can decide whether to merge or not.
